lighttpd 1.4.56

gstrauss opened the pull request:
lighttpd 1.4.56

This depends on #120 

Submitting for review.

lighttpd-1.4.56~rc2 is available now, but lighttpd-1.4.56 has not yet been released.

gstrauss commented:
lighttpd 1.4.57 has been released.  Updated PR to update lighttpd version.

gstrauss commented:
lighttpd 1.4.58 has been released. Updated PR to update lighttpd version.

gstrauss commented:
lighttpd 1.4.59 has been released. Updated PR to update lighttpd version.

gene1wood commented on src/templates/partials/lighttpd.hbs:
```diff
@@ -11,36 +17,87 @@ $SERVER["socket"] == ":80" {
+$SERVER["socket"] ==     ":443" { ssl.engine = "enable" }
+$SERVER["socket"] == "[::]:443" { ssl.engine = "enable" }
+ssl.privkey = "/path/to/private_key"
+ssl.pemfile = "/path/to/signed_cert_plus_intermediates"
```
Now that the pemfile contains intermediates, what would you think about clarifying the order with something like this
```suggestion
ssl.pemfile = "/path/to/signed_cert_followed_by_intermediates"
```

gstrauss replied:
ok.  ackn

gstrauss replied:
PR updated

gene1wood commented on src/templates/partials/lighttpd.hbs:
```diff
@@ -11,36 +17,87 @@ $SERVER["socket"] == ":80" {
+   {{/if}}
+# OCSP stapling (input file must be maintained by external script)
+# https://redmine.lighttpd.net/projects/lighttpd/wiki/Docs_SSL#OCSP-Stapling
+#ssl.stapling-file = "/path/to/cert-staple.der"
```
You probably want to uncomment this line.
```suggestion
ssl.stapling-file = "/path/to/cert-staple.der"
```

gstrauss replied:
No.  I think that is is not appropriate to uncomment a line to a bogus file path.

To enable TLS stapling in lighttpd, the admin must provide (and must regularly update) the file containing stapling DER.  To avoid noise in the error log for a missing file, I think this should remain commented out.

gene1wood replied:
It's just that lines that reference bogus file paths are sort of the standard in the SSL config generator. For example the `ssl.privkey`, `ssl.pemfile` and `ssl.ca-file` lines for lighthttpd and most every line in the other servers that points to a file.

I may be misunderstanding what the `ssl.stapling-file` directive is for. How would it differ from those other directives which are set uncommented and point to bogus files?

gstrauss replied:
If you do not provide `ssl.pemfile`, lighttpd is unable to make TLS work.

On the other hand, if you do not provide `ssl.stapling-file`, TLS will work, but without TLS stapling.  TLS stapling is optional AND requires additional work by the admin to set it up and maintain.

gstrauss replied:
OCSP stapling is a new and optional feature in lighttpd 1.4.56 and later.
Instructions how to maintain `ssl.stapling-file` can be found at [lighttpd TLS OCSP stapling](https://redmine.lighttpd.net/projects/lighttpd/wiki/Docs_SSL#OCSP-Stapling)

gstrauss replied:
I modified the template to uncomment `ssl.stapling-file`, but only show it if `form.ocsp` is enabled (the default)

gene1wood commented on src/templates/partials/lighttpd.hbs:
```diff
@@ -11,36 +17,87 @@ $SERVER["socket"] == ":80" {
 
     # {{output.dhCommand}} > /path/to/dhparam
-    ssl.dh-file               = "/path/to/dhparam"
+    ssl.dh-file = "/path/to/dhparam"
```
Is `ssl.dh-file` meant to only be used in versions greater than `1.4.29` and less than`1.4.56`, because currently `ssl.dh-file` isn't present in a `1.4.56` and newer config.

gstrauss replied:
While lighttpd still supports @ssl.dh-file@, the lighttpd 1.4.56 and later, and TLS libraries provide appropriate defaults.  In other words, this config knob should no longer be used by the average user, though it is still available for those with a specific need.

gstrauss commented on src/templates/partials/lighttpd.hbs:
```diff
@@ -11,36 +17,87 @@ $SERVER["socket"] == ":80" {
+
+{{#if (minver "1.4.56" form.serverVersion)}}
+# select one TLS module: "mod_openssl" "mod_mbedtls" "mod_gnutls" "mod_wolfssl" "mod_nss"
+#server.modules += ("mod_openssl")
```
Yes, the config works with the line commented out.  By default, lighttpd will load mod_openssl if ssl.* directives are present, and a TLS modules, e.g. from among those listed, has not been loaded

gstrauss commented on src/templates/partials/lighttpd.hbs:
```diff
@@ -11,36 +17,87 @@ $SERVER["socket"] == ":80" {
+ssl.openssl.ssl-conf-cmd = ("Protocol" => "ALL, -SSLv2, -SSLv3{{#unless (includes "TLSv1" output.protocols)}}, -TLSv1{{/unless}}{{#unless (includes "TLSv1.1" output.protocols)}}, -TLSv1.1{{/unless}}{{#unless (includes "TLSv1.2" output.protocols)}}, -TLSv1.2{{/unless}}")
+  {{/if}}
+ssl.openssl.ssl-conf-cmd += ("Options" => "{{#if output.serverPreferredOrder}}+{{else}}-{{/if}}ServerPreference")
+   {{#if output.ciphers.length}}
```
ok.  ackn

gstrauss replied:
PR updated

gstrauss commented:
I rebased the PR on master and, per your feedback, added commits to adjust condition indenting and update a label.

The only outstanding discussion is `ssl.stapling-file`.  Please review the one-paragraph doc for [lighttpd TLS OCSP stapling](https://redmine.lighttpd.net/projects/lighttpd/wiki/Docs_SSL#OCSP-Stapling).  lighttpd asks the admin to handle of OCSP stapling externally to lighttpd.  This differs from some other web servers, e.g. nginx, but does not suffer the initial failure of nginx to provide stapling on the the first request after nginx is started.

gstrauss commented:
I modified the template to uncomment `ssl.stapling-file`, but only show it if `form.ocsp` is enabled (the default)

gene1wood reviewed (approved):
Glenn, thanks very much for all your work on this PR. Looks great to me.

gstrauss commented:
@gene1wood: Thank you for your suggestions and feedback.
